src/stan/services/pathfinder/single.hpp

```diff
@@ -0,0 +1,1104 @@
+  const Eigen::Index num_unconstrained_params = names.size() - 2;
+  if (remaining_draws > 0) {
+    try {
+      internal::elbo_est_t est_draws = internal::est_approx_draws<false>(
+          lp_fun, constrain_fun, rng, taylor_approx_best, remaining_draws,
+          taylor_approx_best.alpha, path_num, logger);
+      num_evals += est_draws.fn_calls;
+      auto&& new_lp_ratio = est_draws.lp_ratio;
+      auto&& lp_draws = est_draws.lp_mat;
+      auto&& new_draws = est_draws.repeat_draws;
+      lp_ratio = Eigen::Array<double, -1, 1>(new_lp_ratio.size()
+                                             + elbo_lp_ratio.size());
+      lp_ratio.head(elbo_lp_ratio.size()) = elbo_lp_ratio.array();
+      lp_ratio.tail(new_lp_ratio.size()) = new_lp_ratio.array();
+      const auto total_size = elbo_draws.cols() + new_draws.cols();
+      constrained_draws_mat
+          = Eigen::Array<double, -1, -1>(names.size(), total_size);
+      Eigen::VectorXd unconstrained_col;
+      Eigen::VectorXd approx_samples_constrained_col;
+      for (Eigen::Index i = 0; i < elbo_draws.cols(); ++i) {
+        unconstrained_col = elbo_draws.col(i);
+        constrained_draws_mat.col(i).head(num_unconstrained_params)
+            = constrain_fun(rng, unconstrained_col,
+                            approx_samples_constrained_col);
+        constrained_draws_mat.col(i).tail(2) = elbo_lp_mat.row(i);
+      }
+      for (Eigen::Index i = elbo_draws.cols(), j = 0; i < total_size;
+           ++i, ++j) {
+        unconstrained_col = new_draws.col(j);
+        constrained_draws_mat.col(i).head(num_unconstrained_params)
+            = constrain_fun(rng, unconstrained_col,
+                            approx_samples_constrained_col);
+        constrained_draws_mat.col(i).tail(2) = lp_draws.row(j);
+      }
+    } catch (const std::exception& e) {
+      std::string err_msg = e.what();
+      logger.info(path_num + "Final sampling approximation failed with error: "
+                  + err_msg);
+      logger.info(
+          path_num
+          + "Returning the approximate samples used for ELBO calculation: "
+          + err_msg);
+      constrained_draws_mat
+          = Eigen::Array<double, -1, -1>(names.size(), elbo_draws.cols());
+      Eigen::VectorXd approx_samples_constrained_col;
+      Eigen::VectorXd unconstrained_col;
+      for (Eigen::Index i = 0; i < elbo_draws.cols(); ++i) {
+        unconstrained_col = elbo_draws.col(i);
+        constrained_draws_mat.col(i).head(num_unconstrained_params)
+            = constrain_fun(rng, unconstrained_col,
+                            approx_samples_constrained_col);
+        constrained_draws_mat.col(i).tail(2) = elbo_lp_mat.row(i);
+      }
+      lp_ratio = std::move(elbo_best.lp_ratio);
+    }
+  } else {
+    constrained_draws_mat
+        = Eigen::Array<double, -1, -1>(names.size(), elbo_draws.cols());
+    Eigen::VectorXd approx_samples_constrained_col;
+    Eigen::VectorXd unconstrained_col;
```

Comment:
This declaration of `unconstrained_col` should just move inside the loop as the value's not used outside and the storage isn't re-used.  I'd be tempted to just use `elbo_draws.col(i)` directly on line 1078 and avoid the temporary variable altogether (with a `.eval()` if necessary).

Reply:
Sadly we cannot use `elbo_draws.col(i)` as the model's virtual function does not accept views into a matrix as the input to `log_prob`. Though it would be very nice to add this

Reply:
> This declaration of unconstrained_col should just move inside the loop as the value's not used outside and the storage isn't re-used.

It is reused since `unconstrained_col = elbo_draws.col(i);` makes a hard copy